src/natcap/invest/seasonal_water_yield/seasonal_water_yield.py

```diff
@@ -879,133 +875,178 @@ def _calculate_annual_qfi(qfm_path_list, target_qf_path):
+
+        # qf_im is the quickflow at pixel i on month m
+        qf_im = numpy.full(p_im.shape, TARGET_NODATA, dtype=numpy.float32)
+
+        # case 1: there is no precipitation where both p_im and n_m are
+        # defined and equal to zero.
+        case_1_mask = ~precip_mask
+        qf_im[case_1_mask] = 0
+
+        # case 2: there is precipitation where both p_im and n_m are
+        # defined and greater than zero.
+        case_2_mask = precip_mask & stream_mask
+        qf_im[case_2_mask] = p_im[case_2_mask]
+
+        # case 3
+        # only where all four inputs are defined
+        case_3_mask = valid_mask & precip_mask & ~stream_mask
+
+        term_result = numpy.full(
+            qf_im[case_3_mask].shape, TARGET_NODATA, dtype=numpy.float32)
 
         # a_im is the mean rain depth on a rainy day at pixel i on month m
         # the 25.4 converts inches to mm since Si is in inches
-        a_im = numpy.empty(valid_n_events.shape)
-        a_im = p_im[valid_mask] / (valid_n_events * 25.4)
-        qf_im = numpy.empty(p_im.shape)
-        qf_im[:] = qf_nodata
-
-        # Precompute the last two terms in quickflow so we can handle a
-        # numerical instability when s_i is large and/or a_im is small
-        # on large valid_si/a_im this number will be zero and the latter
-        # exponent will also be zero because of a divide by zero. rather than
-        # raise that numerical warning, just handle it manually
-        E1 = scipy.special.expn(1, valid_si / a_im)
-        E1[valid_si == 0] = 0
-        nonzero_e1_mask = E1 != 0
-        exp_result = numpy.zeros(valid_si.shape)
-        exp_result[nonzero_e1_mask] = numpy.exp(
-            (0.8 * valid_si[nonzero_e1_mask]) / a_im[nonzero_e1_mask] +
-            numpy.log(E1[nonzero_e1_mask]))
-
-        # qf_im is the quickflow at pixel i on month m Eq. [1]
-        qf_im[valid_mask] = (25.4 * valid_n_events * (
-            (a_im - valid_si) * numpy.exp(-0.2 * valid_si / a_im) +
-            valid_si ** 2 / a_im * exp_result))
-
-        # if precip is 0, then QF should be zero
-        qf_im[(p_im == 0) | (n_events == 0)] = 0.0
-        # if we're on a stream, set quickflow to the precipitation
-        valid_stream_precip_mask = stream_array == 1
-        if p_nodata is not None:
-            valid_stream_precip_mask &= ~utils.array_equals_nodata(
-                p_im, p_nodata)
-        qf_im[valid_stream_precip_mask] = p_im[valid_stream_precip_mask]
+        a_im = p_im[case_3_mask] / (n_m[case_3_mask] * 25.4)
+
+        # don't raise a warning on overflow
+        with numpy.errstate(over='ignore'):
+            exp_result = numpy.exp((0.8 * s_i[case_3_mask]) / a_im)
+
+        # edge case 3a: set the whole term to 0 when s_i = 0
+        subcase_a_mask = s_i[case_3_mask] == 0
+        term_result[subcase_a_mask] = 0
+
+        # edge case 3b: set the whole term to 0 when exp() overflows
+        subcase_b_mask = numpy.isinf(exp_result)
```

Comment:
Interesting, so ignoring the `OverflowError` will leave `exp_result` with an infinity value?

Reply:
Yes:
```
>>> numpy.exp(800)
<stdin>:1: RuntimeWarning: overflow encountered in exp
inf
```
To your question above
> is there a reason you chose to "ignore" the error vs surround it in a try / except block?

Because it's a warning not an error, and a RuntimeWarning could be caused by other things whereas the numpy context manager lets you specifically ignore overflows

Reply:
Oh interesting! I was testing `math.exp(800)` and getting:
```python
>>> math.exp(900)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: math range error
```
That makes sense.